build: add pkg-config support

Tachi107 opened the pull request:
Fixes #1763

I chose the "dynamic" route :)
It is a bit more complex than a more static solution, but it has the advantage of requiring just two lines of code (module inclusion and function call) in "user" code and should also work in every project. Of course this is not perfect (for example, it doesn't currently handle cases like generating a .pc file of a library that links to an imported target that has no imported location but only a list of more than one interface libraries), but works in most cases. Definitely works in a project like simdjson that doesn't have external dependencies ;)

Edit: oh right, tests... To test this you would need to test the library as installed - compiling the library, installing it, and compiling a sample program that links to this using pkg-config. I don't know if you already have a setup like this in your CI, but it is basically what Debian's autopkgtest ([DEP-8](https://dep-team.pages.debian.net/deps/dep8/)) does

Tachi107 commented:
I'd like to note that currently the generated pkg-config file sets the `Version` field to `9.0.0` because `SIMDJSON_LIB_VERSION` is set to that.

lemire commented:
> To test this you would need to test the library as installed - compiling the library, installing it, and compiling a sample program that links to this using pkg-config. I don't know if you already have a setup like this in your CI

We do. Yes. We use `cmake install`, see 

https://github.com/simdjson/simdjson/blob/master/.github/workflows/ubuntu20.yml

For testing purposes, it is not always handy to install the package system-wide (this makes local tests annoying and local tests are important if one wants to debug the issue). So we have `cmake install` go to a custom (non system) directory. Thankfully, it appears that pkg-config supports that as well: `PKG_CONFIG_PATH`.

lemire commented:
I restarted circle-ci.

lemire commented:
I imagine we would like to be able to do something like...

```
cmake  -DCMAKE_INSTALL_PREFIX:PATH=destination -B build
cmake --build build
cmake --install build
PKG_CONFIG_PATH=destination ....
```

lemire commented:
There are errors currently such as...

```

Threads::Threads is imported, and has -pthread
CMake Error at cmake/gen_pkg_config.cmake:77 (get_target_property):
  INTERFACE_LIBRARY targets may only have whitelisted properties.  The
  property "IMPORTED_LOCATION" is not allowed.
Call Stack (most recent call first):
  CMakeLists.txt:165 (gen_pkg_config)
```

Tachi107 commented:
> I imagine we would like to be able to do something like...
> 
> ```
> cmake  -DCMAKE_INSTALL_PREFIX:PATH=destination -B build
> cmake --build build
> cmake --install build
> PKG_CONFIG_PATH=destination ....
> ```
> 

I don't understand what you mean; you'd like to install the .pc in a custom directory? Those files have to be installed in libdir/pkgconfig, I don't think that changing the directory is allowed/useful (just like cmake config files, that have to be installed in libdir/cmake)

As for the errors, thanks for reporting. I didn't encounter them, but I definitely have to improve this module. I'll mark this as a draft for the time being.

lemire commented:
> I don't understand what you mean; you'd like to install the .pc in a custom directory? Those files have to be installed in libdir/pkgconfig, I don't think that changing the directory is allowed/useful (just like cmake config files, that have to be installed in libdir/cmake)

The cmake config files do not have to be installed system wide. You can set `CMAKE_PREFIX_PATH`.

Here is what the `pkg-config` man page says...

_By default, pkg-config looks in the directory prefix/lib/pkgconfig for these files; it will also look in the colon-separated (on Windows, semicolon-separated) list of directories specified by the PKG_CONFIG_PATH environment variable._

It is useful to various reasons:
1. When you want to test and debug, you typically do not want to install the files system wide. It is much cleaner to install them in an isolated directory that you can delete as needed.
2. Not all users want to install a library in the default locations. For example, it is perfectly reasonable to have a user-specific installation. There is no need to make a library available system-wide when a single user needs it.

lemire commented:
> The cmake config files do not have to be installed system wide. You can set CMAKE_PREFIX_PATH.

In fact, if you follow my link above, you will find that this is exactly what we do currently in CI. It is relatively simple: cmake will use this variable both for installing and searching packages. So you can install cmake libraries wherever you'd like on a system.

Tachi107 commented:
This should honor CMAKE_PREFIX_PATH too

Tachi107 commented:
Ok, I now understand what you meant- sorry but I wasn't at home and I replied without reading carefully what you wrote. The .pc file generated with this module should already get installed relative to `CMAKE_PREFIX_PATH`, and invoking `pkg-config` (or `pkgconf`) with `PKG_CONFIG_PATH` set to `$CMAKE_PREFIX_PATH/lib/pkgconfig` should work as expected

Tachi107 commented:
I gave up implementing a full pkg-config generator from scratch- it really should be implemented in upstream CMake.

On the other hand, I finalized "classic" pkg-config generation, i.e. the one that uses a .pc.in template file.

You can read the updated description in the commit message, but I'll include it here for convenience:

> The CMake build script now generates a simple pkg-config files that can be easily used by non-CMake users.
> 
> The file is generated from a template file that gets filled in at configure time.
> 
> As CMake doesn't have anything similar to Meson's pkg-config generator the file is quite static, i.e. new simdjson public defines/dependencies won't be picked up automatically.
> 
> This approach also suffers from one minor issue, mentioned in [jtojnar/cmake-snips]; in short, it doesn't work well when users specify CMAKE_INSTALL_INCLUDEDIR and similar as absolute paths. It's not a big deal, and it will easily fixable once you'll require CMake >=3.20.
> 
> Fixes #1763
> 
> [jtojnar/cmake-snips]: https://github.com/jtojnar/cmake-snips#concatenating-paths-when-building-pkg-config-files

eli-schwartz commented on simdjson.pc.in:
```diff
@@ -0,0 +1,11 @@
+prefix=@CMAKE_INSTALL_PREFIX@
+includedir=${prefix}/@CMAKE_INSTALL_INCLUDEDIR@
+libdir=${prefix}/@CMAKE_INSTALL_LIBDIR@
```
Why not just use the `JoinPaths` module anyway, for now, so that you properly handle the case where CMAKE_INSTALL_LIBDIR is an absolute path?

Tachi107 replied:
Mainly because it seems that it doesn't work well on Windows, but yeah it's better than nothing. I'll add it later.

Tachi107 commented:
Sorry, I completely forgot about this PR! I've now addressed absolute path handling, and this should be now ready to be merged.

I've added the fix in a separate commit so that it'll be easy to revert it, just in case.

Tachi107 commented:
I've also added a CI job that tests the generated pkg-config file. I didn't know where to add it, so I just created a new Debian Testing runner, partly because it seems that you don't have a rolling-release CI.

Tachi107 commented:
@lemire sorry for the ping, but this is now ready and tested by CI. Should I do something else to get this merged?

lemire commented:
Merging.

Tachi107 commented:
Thanks :tada:
